opm/ml/keras_model.cpp

```diff
@@ -0,0 +1,419 @@
+            "Expected biases");
+
+    KASSERT(activation_.LoadLayer(file), "Failed to load activation");
+
+    return true;
+}
+
+template<class Evaluation>
+bool KerasLayerDense<Evaluation>::Apply(Tensor<Evaluation>* in, Tensor<Evaluation>* out) {
+    KASSERT(in, "Invalid input");
+    KASSERT(out, "Invalid output");
+    KASSERT(in->dims_.size() <= 2, "Invalid input dimensions");
+
+    if (in->dims_.size() == 2) {
+        KASSERT(in->dims_[1] == weights_.dims_[0], "Dimension mismatch %d %d",
+                in->dims_[1], weights_.dims_[0]);
+    }
+
+    Tensor<Evaluation> tmp(weights_.dims_[1]);
+
+    for (int i = 0; i < weights_.dims_[0]; i++) {
+        for (int j = 0; j < weights_.dims_[1]; j++) {
+            tmp(j) += (*in)(i)*weights_(i, j);
+        }
+    }
+
+    for (int i = 0; i < biases_.dims_[0]; i++) {
+        tmp(i) += biases_(i);
+    }
+
+    KASSERT(activation_.Apply(&tmp, out), "Failed to apply activation");
+
+    return true;
+}
+
+template<class Evaluation>
+bool KerasLayerFlatten<Evaluation>::LoadLayer(std::ifstream* file) {
+    KASSERT(file, "Invalid file stream");
+    return true;
+}
+
+template<class Evaluation>
+bool KerasLayerFlatten<Evaluation>::Apply(Tensor<Evaluation>* in, Tensor<Evaluation>* out) {
+    KASSERT(in, "Invalid input");
+    KASSERT(out, "Invalid output");
+
+    *out = *in;
+    out->Flatten();
+
+    return true;
+}
+
+
+
+template<class Evaluation>
+bool KerasLayerEmbedding<Evaluation>::LoadLayer(std::ifstream* file) {
+    KASSERT(file, "Invalid file stream");
+
+    unsigned int weights_rows = 0;
+    KASSERT(ReadUnsignedInt(file, &weights_rows), "Expected weight rows");
+    KASSERT(weights_rows > 0, "Invalid weights # rows");
+
+    unsigned int weights_cols = 0;
+    KASSERT(ReadUnsignedInt(file, &weights_cols), "Expected weight cols");
+    KASSERT(weights_cols > 0, "Invalid weights shape");
+
+    weights_.Resize(weights_rows, weights_cols);
+    KASSERT(
+        ReadFloats(file, weights_.data_.data(), weights_rows * weights_cols),
+        "Expected weights");
+
+    return true;
+}
+
+template<class Evaluation>
+bool KerasLayerEmbedding<Evaluation>::Apply(Tensor<Evaluation>* in, Tensor<Evaluation>* out) {
+    int output_rows = in->dims_[1];
+    int output_cols = weights_.dims_[1];
+    out->dims_ = {output_rows, output_cols};
+    out->data_.reserve(output_rows * output_cols);
+
+    std::for_each(in->data_.begin(), in->data_.end(), [=](Evaluation i) {
+        typename std::vector<Evaluation>::const_iterator first =
+            this->weights_.data_.begin() + (getValue(i) * output_cols);
+        typename std::vector<Evaluation>::const_iterator last =
+            this->weights_.data_.begin() + (getValue(i) + 1) * output_cols;
+
+        out->data_.insert(out->data_.end(), first, last);
+    });
+
+    return true;
+}
+
+
+template<class Evaluation>
+bool KerasModel<Evaluation>::LoadModel(const std::string& filename) {
```

Comment:
I assume this file format is defined by Keras? Please include in a comment a reference.